File: Tests.VisualBasic/Stringier/Encodings/UTF8Tests.vb
Imports Generators
Imports System
Imports System.Diagnostics.CodeAnalysis
Imports System.Text
Imports Stringier.Encodings
Imports Xunit

Namespace Stringier.Encodings
    Public Class UTF8Tests
        <Theory>
        <InlineData(&H0, &H0)>
        <InlineData(&H41, &H41)>
        <InlineData(&H7F, &H7F)>
        <InlineData(&H80, &HFFFD)>
        <InlineData(&HFF, &HFFFD)>
        Public Sub Decode_1(ByVal first As Byte, ByVal expected As Int32)
            Assert.Equal(expected, UTF8.Decode(first).Value)
        End Sub

        <Theory>
        <InlineData(&HC3, &H86, &HC6)>
        <InlineData(&HCE, &HB2, &H3B2)>
        Public Sub Decode_2(ByVal first As Byte, ByVal second As Byte, ByVal expected As Int32)
            Assert.Equal(expected, UTF8.Decode(first, second).Value)
        End Sub

        <Theory>
        <InlineData(&HE0, &HAE, &H87, &HB87)>
        <InlineData(&HE1, &H82, &HA7, &H10A7)>
        <InlineData(&HEA, &H9C, &HA8, &HA728)>
        Public Sub Decode_3(ByVal first As Byte, ByVal second As Byte, ByVal third As Byte, ByVal expected As Int32)
            Assert.Equal(expected, UTF8.Decode(first, second, third).Value)
        End Sub

        <Theory>
        <InlineData(&HF0, &H90, &H8C, &HB1, &H10331)>
        <InlineData(&HF0, &H9F, &H82, &HA1, &H1F0A1)>
        Public Sub Decode_4(ByVal first As Byte, ByVal second As Byte, ByVal third As Byte, ByVal fourth As Byte, ByVal expected As Int32)
            Assert.Equal(expected, UTF8.Decode(first, second, third, fourth).Value)
        End Sub

        <Theory>
        <InlineData(New Byte() {}, New Int32() {})>
        <InlineData(New Byte() {&H0}, New Int32() {&H0})>
        <InlineData(New Byte() {&H41, &H7F}, New Int32() {&H41, &H7F})>
        <InlineData(New Byte() {&H41, &HC3, &H86}, New Int32() {&H41, &HC6})>
        <InlineData(New Byte() {&HC3, &H86, &H41}, New Int32() {&HC6, &H41})>
        <InlineData(New Byte() {&HE0, &HAE, &H87}, New Int32() {&HB87})>
        <InlineData(New Byte() {&HE0, &HAE, &HE0}, New Int32() {&HFFFD})>
        Public Sub Decode_Buffer(ByVal bytes As Byte(), ByVal expected As Int32())
            Dim text As Rune() = New Rune(expected.Length - 1) {}

            For i As Int32 = 0 To text.Length - 1
                text(i) = New Rune(expected(i))
            Next

            Assert.Equal(text, UTF8.Decode(bytes).ToArray())
        End Sub

        <Theory>
        <InlineData(4096)>
        <InlineData(2048)>
        <InlineData(1024)>
        <InlineData(512)>
        <InlineData(256)>
        <InlineData(128)>
        <SuppressMessage("Blocker Code Smell", "S2699:Tests should include assertions", Justification:="We're just trying to make sure this doesn't throw an exception, which would result in a failed test anyways.")>
        Public Sub Decode_Gibberish(ByVal reductionFactor As Int32)
            Dim buffer As Byte() = Gibberish.GenerateUTF8(reductionFactor)
            UTF8.Decode(buffer)
        End Sub

        <Theory>
        <InlineData(&H41, New Byte() {&H41})>
        <InlineData(&HC6, New Byte() {&HC3, &H86})>
        <InlineData(&HB87, New Byte() {&HE0, &HAE, &H87})>
        <InlineData(&HA728, New Byte() {&HEA, &H9C, &HA8})>
        <InlineData(&H10331, New Byte() {&HF0, &H90, &H8C, &HB1})>
        <InlineData(&H1F0A1, New Byte() {&HF0, &H9F, &H82, &HA1})>
        Public Sub Encode(ByVal scalarValue As UInt32, ByVal expected As Byte())
            Assert.Equal(expected, UTF8.Encode(New Rune(scalarValue)).ToArray())
        End Sub

        <Theory>
        <InlineData(4096)>
        <InlineData(2048)>
        <InlineData(1024)>
        <InlineData(512)>
        <InlineData(256)>
        <InlineData(128)>
        <SuppressMessage("Blocker Code Smell", "S2699:Tests should include assertions", Justification:="We're just trying to make sure this doesn't throw an exception, which would result in a failed test anyways.")>
        Public Sub Encode_Gibberish(ByVal reductionFactor As Int32)
            Dim buffer As Rune() = Gibberish.GenerateUTF32(reductionFactor)
            UTF8.Encode(buffer)
        End Sub

        <Theory>
        <InlineData(&H0, True)>
        <InlineData(&H1, True)>
        <InlineData(&H7E, True)>
        <InlineData(&H7F, True)>
        <InlineData(&H80, False)>
        <InlineData(&H81, False)>
        <InlineData(&HC0, False)>
        <InlineData(&HC1, False)>
        <InlineData(&HC2, True)>
        <InlineData(&HF4, True)>
        <InlineData(&HF5, False)>
        <InlineData(&HF6, False)>
        <InlineData(&HFF, False)>
        Public Sub IsFirstByte(ByVal [Byte] As Byte, ByVal expected As Boolean)
            Assert.Equal(expected, UTF8.IsFirstByte([Byte]))
        End Sub

        <Theory>
        <InlineData(&H7F, &H80, &HBF)>
        <InlineData(&HDF, &H80, &HBF)>
        <InlineData(&HE0, &HA0, &HBF)>
        <InlineData(&HE1, &H80, &HBF)>
        <InlineData(&HEE, &H80, &HBF)>
        <InlineData(&HED, &H80, &H9F)>
        <InlineData(&HEF, &H80, &HBF)>
        <InlineData(&HF0, &H90, &HBF)>
        <InlineData(&HF1, &H80, &HBF)>
        <InlineData(&HF3, &H80, &HBF)>
        <InlineData(&HF4, &H80, &H8F)>
        <InlineData(&HF5, &H80, &HBF)>
        <InlineData(&HFF, &H80, &HBF)>
        Public Sub NextByteRange(ByVal [Byte] As Byte, ByVal lowerExp As Byte, ByVal upperExp As Byte)
            Dim lower As Byte = Nothing, upper As Byte = Nothing
            UTF8.NextByteRange([Byte], lower, upper)
            Assert.Equal(lowerExp, lower)
            Assert.Equal(upperExp, upper)
        End Sub

        <Theory>
        <InlineData(&H0, 1)>
        <InlineData(&H7F, 1)>
        <InlineData(&H80, 0)>
        <InlineData(&HC1, 0)>
        <InlineData(&HC2, 2)>
        <InlineData(&HDF, 2)>
        <InlineData(&HE0, 3)>
        <InlineData(&HE1, 3)>
        <InlineData(&HEE, 3)>
        <InlineData(&HEF, 3)>
        <InlineData(&HF0, 4)>
        <InlineData(&HF1, 4)>
        <InlineData(&HF3, 4)>
        <InlineData(&HF4, 4)>
        <InlineData(&HF5, 0)>
        <InlineData(&HF6, 0)>
        <InlineData(&HFF, 0)>
        Public Sub SequenceLength(ByVal [Byte] As Byte, ByVal expected As Int32)
            Assert.Equal(expected, UTF8.SequenceLength([Byte]))
        End Sub
    End Class
End Namespace

File: Tests/Stringier/Encodings/UTF8Tests.cs
using Generators;
using System;
using System.Diagnostics.CodeAnalysis;
using System.Text;
using Xunit;

namespace Stringier.Encodings {
	public class UTF8Tests {
		[Theory]
		[InlineData(0x0, 0x0)]
		[InlineData(0x41, 0x41)]
		[InlineData(0x7F, 0x7F)]
		[InlineData(0x80, 0xFFFD)]
		[InlineData(0xFF, 0xFFFD)]
		public void Decode_1(Byte first, Int32 expected) => Assert.Equal(expected, UTF8.Decode(first).Value);

		[Theory]
		[InlineData(0xC3, 0x86, 0xC6)]
		[InlineData(0xCE, 0xB2, 0x03B2)]
		public void Decode_2(Byte first, Byte second, Int32 expected) => Assert.Equal(expected, UTF8.Decode(first, second).Value);

		[Theory]
		[InlineData(0xE0, 0xAE, 0x87, 0x0B87)]
		[InlineData(0xE1, 0x82, 0xA7, 0x10A7)]
		[InlineData(0xEA, 0x9C, 0xA8, 0xA728)]
		public void Decode_3(Byte first, Byte second, Byte third, Int32 expected) => Assert.Equal(expected, UTF8.Decode(first, second, third).Value);

		[Theory]
		[InlineData(0xF0, 0x90, 0x8C, 0xB1, 0x010331)]
		[InlineData(0xF0, 0x9F, 0x82, 0xA1, 0x01F0A1)]
		public void Decode_4(Byte first, Byte second, Byte third, Byte fourth, Int32 expected) => Assert.Equal(expected, UTF8.Decode(first, second, third, fourth).Value);

		[Theory]
		[InlineData(new Byte[0], new Int32[0])]
		[InlineData(new Byte[] { 0x0 }, new Int32[] { 0x0 })]
		[InlineData(new Byte[] { 0x41, 0x7F }, new Int32[] { 0x41, 0x7F })]
		[InlineData(new Byte[] { 0x41, 0xC3, 0x86 }, new Int32[] { 0x41, 0xC6 })]
		[InlineData(new Byte[] { 0xC3, 0x86, 0x41 }, new Int32[] { 0xC6, 0x41 })]
		[InlineData(new Byte[] { 0xE0, 0xAE, 0x87 }, new Int32[] { 0x0B87 })]
		[InlineData(new Byte[] { 0xE0, 0xAE, 0xE0 }, new Int32[] { 0xFFFD })]
		public void Decode_Buffer(Byte[] buffer, Int32[] expected) {
			Rune[] text = new Rune[expected.Length];
			for (Int32 i = 0; i < text.Length; i++) {
				text[i] = new Rune(expected[i]);
			}
			Assert.Equal(text, UTF8.Decode(buffer).ToArray());
		}

		[Theory]
		[InlineData(128)]
		[InlineData(256)]
		[InlineData(512)]
		[InlineData(1024)]
		[InlineData(2048)]
		[InlineData(4096)]
		[SuppressMessage("Blocker Code Smell", "S2699:Tests should include assertions", Justification = "We're just trying to make sure this doesn't throw an exception, which would result in a failed test anyways.")]
		public void Decode_Gibberish(Int32 size) => UTF8.Decode(Gibberish.GenerateUTF8(size));

		[Theory]
		[InlineData(0x000041, new Byte[] { 0x41 })]
		[InlineData(0x0000C6, new Byte[] { 0xC3, 0x86 })]
		[InlineData(0x000B87, new Byte[] { 0xE0, 0xAE, 0x87 })]
		[InlineData(0x00A728, new Byte[] { 0xEA, 0x9C, 0xA8 })]
		[InlineData(0x010331, new Byte[] { 0xF0, 0x90, 0x8C, 0xB1 })]
		[InlineData(0x01F0A1, new Byte[] { 0xF0, 0x9F, 0x82, 0xA1 })]
		public void Encode(UInt32 scalarValue, Byte[] expected) {
			Assert.Equal(expected, UTF8.Encode(new Rune(scalarValue)).ToArray());
		}

		[Theory]
		[InlineData(128)]
		[InlineData(256)]
		[InlineData(512)]
		[InlineData(1024)]
		[InlineData(2048)]
		[InlineData(4096)]
		[SuppressMessage("Blocker Code Smell", "S2699:Tests should include assertions", Justification = "We're just trying to make sure this doesn't throw an exception, which would result in a failed test anyways.")]
		public void Encode_Gibberish(Int32 size) => UTF8.Encode(Gibberish.GenerateUTF32(size));

		[Theory]
		[InlineData(0x00, true)]
		[InlineData(0x01, true)]
		[InlineData(0x7E, true)]
		[InlineData(0x7F, true)]
		[InlineData(0x80, false)]
		[InlineData(0x81, false)]
		[InlineData(0xC0, false)]
		[InlineData(0xC1, false)]
		[InlineData(0xC2, true)]
		[InlineData(0xF4, true)]
		[InlineData(0xF5, false)]
		[InlineData(0xF6, false)]
		[InlineData(0xFF, false)]
		public void IsFirstByte(Byte @byte, Boolean expected) => Assert.Equal(expected, UTF8.IsFirstByte(@byte));

		[Theory]
		[InlineData(0x7F, 0x80, 0xBF)]
		[InlineData(0xDF, 0x80, 0xBF)]
		[InlineData(0xE0, 0xA0, 0xBF)]
		[InlineData(0xE1, 0x80, 0xBF)]
		[InlineData(0xEE, 0x80, 0xBF)]
		[InlineData(0xED, 0x80, 0x9F)]
		[InlineData(0xEF, 0x80, 0xBF)]
		[InlineData(0xF0, 0x90, 0xBF)]
		[InlineData(0xF1, 0x80, 0xBF)]
		[InlineData(0xF3, 0x80, 0xBF)]
		[InlineData(0xF4, 0x80, 0x8F)]
		[InlineData(0xF5, 0x80, 0xBF)]
		[InlineData(0xFF, 0x80, 0xBF)]
		public void NextByteRange(Byte @byte, Byte lowerExp, Byte upperExp) {
			UTF8.NextByteRange(@byte, out Byte lower, out Byte upper);
			Assert.Equal(lowerExp, lower);
			Assert.Equal(upperExp, upper);
		}

		[Theory]
		[InlineData(0x00, 1)]
		[InlineData(0x7F, 1)]
		[InlineData(0x80, 0)]
		[InlineData(0xC1, 0)]
		[InlineData(0xC2, 2)]
		[InlineData(0xDF, 2)]
		[InlineData(0xE0, 3)]
		[InlineData(0xE1, 3)]
		[InlineData(0xEE, 3)]
		[InlineData(0xEF, 3)]
		[InlineData(0xF0, 4)]
		[InlineData(0xF1, 4)]
		[InlineData(0xF3, 4)]
		[InlineData(0xF4, 4)]
		[InlineData(0xF5, 0)]
		[InlineData(0xF6, 0)]
		[InlineData(0xFF, 0)]
		public void SequenceLength(Byte @byte, Int32 expected) => Assert.Equal(expected, UTF8.SequenceLength(@byte));
	}
}
